graph/graph_cache.go

```diff
@@ -0,0 +1,376 @@
+package graph
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/kiali/kiali/config"
+	"github.com/kiali/kiali/log"
+	"github.com/kiali/kiali/prometheus/internalmetrics"
+)
+
+// GraphCache provides per-session graph caching with background refresh.
+// Each session's graph is cached and refreshed in the background
+// with a moving time window to ensure users always see current traffic data.
+// Sessions are uniquely identified by sessionID (stored in browser cookies).
+// Multiple tabs in the same browser share the same session and cache.
+// Different browsers or incognito windows have separate sessions.
+type GraphCache interface {
+	// ActiveSessions returns the number of sessions with cached graphs
+	ActiveSessions() int
+
+	// Clear removes all cached graphs (useful for testing and shutdown)
+	Clear()
+
+	// Config returns the cache configuration
+	Config() *GraphCacheConfig
+
+	// Enabled returns true if graph caching is enabled
+	Enabled() bool
+
+	// Evict removes a session's graph from cache and stops background refresh
+	Evict(sessionID string)
+
+	// GetGraphGenerator returns the current graph generator (may be nil)
+	GetGraphGenerator() GraphGenerator
+
+	// GetSessionGraph retrieves a session's cached graph if it exists
+	GetSessionGraph(sessionID string) (*CachedGraph, bool)
+
+	// SetGraphGenerator sets the graph generator function for background refresh
+	SetGraphGenerator(generator GraphGenerator)
+
+	// SetSessionGraph stores or updates a session's cached graph
+	SetSessionGraph(sessionID string, cached *CachedGraph) error
+
+	// TotalMemoryMB returns estimated total memory usage of all cached graphs
+	TotalMemoryMB() float64
+}
```

Comment:
If there's not a second implementation or a need to mock out the graph `GraphCache` then having an interface can just add boilerplate with all the getter/setters etc.

Reply:
True. But leaving as-is.